pkg/mig/config/config.go

```diff
@@ -211,76 +215,117 @@ func (m *nvmlMigConfigManager) SetMigConfig(gpu int, config types.MigConfig) err
 	mode, _, ret := device.GetMigMode()
 	if ret.Value() == nvml.ERROR_NOT_SUPPORTED {
-		return fmt.Errorf("MIG not supported")
+		return desiredConfig, fmt.Errorf("MIG not supported")
 	}
 	if ret.Value() != nvml.SUCCESS {
-		return fmt.Errorf("error getting MIG mode: %v", ret)
+		return desiredConfig, fmt.Errorf("error getting MIG mode: %v", ret)
 	}
 	if mode != nvml.DEVICE_MIG_ENABLE {
-		return fmt.Errorf("MIG mode disabled")
+		return desiredConfig, fmt.Errorf("MIG mode disabled")
 	}
 
+	instancesToNotCreate := map[int]bool{}
 	for i := 0; i < nvml.GPU_INSTANCE_PROFILE_COUNT; i++ {
 		giProfileInfo, ret := device.GetGpuInstanceProfileInfo(i)
 		if ret.Value() == nvml.ERROR_NOT_SUPPORTED {
 			continue
 		}
 		if ret.Value() != nvml.SUCCESS {
-			return fmt.Errorf("error getting GPU instance profile info for '%v': %v", i, ret)
+			return desiredConfig, fmt.Errorf("error getting GPU instance profile info for '%v': %v", i, ret)
 		}
 
 		gis, ret := device.GetGpuInstances(&giProfileInfo)
 		if ret.Value() != nvml.SUCCESS {
-			return fmt.Errorf("error getting GPU instances for profile '%v': %v", i, ret)
+			return desiredConfig, fmt.Errorf("error getting GPU instances for profile '%v': %v", i, ret)
 		}
 
 		for _, gi := range gis {
+			destroyGi := true
 			for j := 0; j < nvml.COMPUTE_INSTANCE_PROFILE_COUNT; j++ {
 				for k := 0; k < nvml.COMPUTE_INSTANCE_ENGINE_PROFILE_COUNT; k++ {
 					ciProfileInfo, ret := gi.GetComputeInstanceProfileInfo(j, k)
 					if ret.Value() == nvml.ERROR_NOT_SUPPORTED {
 						continue
 					}
 					if ret.Value() != nvml.SUCCESS {
-						return fmt.Errorf("error getting Compute instance profile info for '(%v, %v)': %v", j, k, ret)
+						return desiredConfig, fmt.Errorf("error getting Compute instance profile info for '(%v, %v)': %v", j, k, ret)
 					}
 
 					cis, ret := gi.GetComputeInstances(&ciProfileInfo)
 					if ret.Value() != nvml.SUCCESS {
-						return fmt.Errorf("error getting Compute instances for profile '(%v, %v)': %v", j, k, ret)
+						return desiredConfig, fmt.Errorf("error getting Compute instances for profile '(%v, %v)': %v", j, k, ret)
 					}
 
 					for _, ci := range cis {
 						ret := ci.Destroy()
 						if ret.Value() != nvml.SUCCESS {
-							return fmt.Errorf("error destroying Compute instance for profile '(%v, %v)': %v", j, k, ret)
+							if ret.Value() == nvml.ERROR_IN_USE && len(desiredConfig) > 0 && destroyGi {
+								gpuInfo, ret := gi.GetInfo()
+								if ret.Value() != nvml.SUCCESS {
+									return desiredConfig, fmt.Errorf("error destroying Compute instance for profile '(%v, %v)': %v", j, k, ret)
+								}
+								index := getMigProfileInConfigByPosition(desiredConfig, device, gpuInfo.ProfileId, instancesToNotCreate)
+								if index != -1 {
+									instancesToNotCreate[index] = true
+									destroyGi = false
+								}
+							} else {
+								return desiredConfig, fmt.Errorf("error destroying Compute instance for profile '(%v, %v)': %v", j, k, ret)
+							}
```

Comment:
This will not work in general, as the physical location of the MIG device plays a role here.

It works in your case of going from `all-balanced` to `all-1g.5gb` with a `1g.5gb` device running because no matter what, all of the new devices were going to be `1g.5gb` anyway.

If however, you were to go the other way around, the `1g.5gb` device that is "in-use" may not be the one that is allocatable in the `all-balanced` configuration.

Reply:
You are totally right. 

In case which the device in used must be removed in order to create other devices it will throw error in the creation stage, and all the configuration process will fail.

But in the case where the device in use is able to stay running and all the other devices are able to be created successfully its cancelling the necessity to stop the running process.

In general, it is not always that when you have a running process on the device it means you have to stop the configuration operation - there are use cases which you can still configure according to the desired config.

I think it's better than stoping the entire configuration process. 
What do you think?

Reply:
Can you explain the use case here a bit better? The current model assumes you *only* reconfigure when a GPU is fully drained and has nothing running on it. I’m not comfortable tweaking that assumption unless I understand why/how you plan on using this.